File: strings/medium/recursion_backtracking/letter_case_permutation_784.go
package recursion_backtracking

func letterCasePermutation(s string) []string {
	result := make([]string, 0)
	backtrackLetterCasePermutation(&result, []rune(s), 0)
	return result
}

func backtrackLetterCasePermutation(result *[]string, s []rune, start int) {
	*result = append(*result, string(s))
	for i := start; i < len(s); i++ {
		isLowerCase := s[i] >= 'a' && s[i] <= 'z'
		isUpperCase := s[i] >= 'A' && s[i] <= 'Z'
		if isLowerCase || isUpperCase {
			if isLowerCase {
				s[i] = s[i] - 32
			}
			if isUpperCase {
				s[i] = s[i] + 32
			}
			backtrackLetterCasePermutation(result, s, i+1)
			if isLowerCase {
				s[i] = s[i] + 32
			}
			if isUpperCase {
				s[i] = s[i] - 32
			}
		} else {
			continue
		}
	}
}

File: leetcode/backtracking_recursion/java_solutions/LetterCasePermutation_784.java
package leetcode.backtracking_recursion.java_solutions;

import java.util.ArrayList;
import java.util.List;

public class LetterCasePermutation_784 {

    public List<String> letterCasePermutation(String s) {
        List<String> result = new ArrayList<>();

        backtrack(result, s.toCharArray(), 0);

        return result;
    }

    private void backtrack(List<String> result, char[] word, int start) {
        result.add(new String(word));
        for (int i = start; i < word.length; i++) {
            if (Character.isLetter(word[i])) {
                boolean isUpperCase = Character.isUpperCase(word[i]);
                boolean isLowerCase = Character.isLowerCase(word[i]);

                if (isUpperCase) {
                    word[i] = Character.toLowerCase(word[i]);
                }
                if (isLowerCase) {
                    word[i] = Character.toUpperCase(word[i]);
                }
                backtrack(result, word, i + 1);
                if (isUpperCase) {
                    word[i] = Character.toUpperCase(word[i]);
                }
                if (isLowerCase) {
                    word[i] = Character.toLowerCase(word[i]);
                }
            }
        }
    }
}
